### ISSUE TITLE
Suggested Improvements to Staging's Implementation of the LS Command

### ISSUE THREAD

**Issue by MasterO2** [dosbox-staging/dosbox-staging#1336]
Looking at the LS command that is now included with Dosbox-Staging, the following switches would also be useful to have included as well if that's possible:

```
-d, --directory  list directories themselves, not their contents
-h, --human-readable
-l, use a long listing format
-R, --recursive  list subdirectories recursively
-X, sort alphabetically by entry extension

https://www.man7.org/linux/man-pages/man1/ls.1.html
```
Also, for large directories that take more than one page to display, LS should be pipe-able to the `more` and `less` commands, since Staging's LS itself does not have the ability to scroll up or down.

**Comment by Grounded0** [comment]
> Also, for large directories that take more than one page to display, LS should be pipe-able to the `more` and `less` commands, since Staging's LS itself does not have the ability to scroll up or down.

It already pipes just fine to more and less last time i checked.

**Comment by brandonrobertz** [comment]
> As i said i have non-standard installation so if someone has energy to check against bog-standard DOSBox Staging install to confirm it actually doesn't work then thank you.

I've confirmed what  MasterO2 is saying on a default dosbox-staging environment. Running FreeDOS is probably getting you a pipe that works more robustly.

**Comment by MasterO2** [comment]
@Wengier In addition to adding the ability to pipe commands to Staging, would you also be willing to add the above improvements to the LS command?

**Comment by Wengier** [comment]
@MasterO2 As a command originally from *nix systems, LS command has a lot of *nix-like options. I hope to clean up and convert some particularly useful ones to DOS style (e.g. ``/d`` instead of ``-d`` or ``--directory``) to work in DOSBox (which is a DOS emulator). But the details can be discussed.

**Comment by Wengier** [comment]
Piping is supported via PR #1517.

**Comment by kcgen** [comment]
Thanks for these additions @Wengier .

Regarding LS, it's only purpose is to keep muscle-memory intact for users primarily coming from Linux / macOS / BSD / MSYS2. They can continue to rapidly type `ls` (and common flags) without having to think twice. 

So we don't want to take away it's handful of existing GNU flags: doing so would confuse that muscle-memory: _"Wait, does this one need -d or /d?"_). 

That said, certainly no harm in adding DOS-style slash handling for the existing GNU-shorthand (single-dash) flags!

**Comment by Wengier** [comment]
@kcgen I certainly understand the purpose you mentioned, since I have similar experiences. For example, Linux provides both ``dir`` and ``ls`` commands, and I personally usually use the former. But in macOS, only ``ls`` command is provided, so I had to always ensure a command like ``alias dir=ls`` is in the shell script (e.g. .bashrc) so that ``dir`` will also work as expected in macOS. I certainly understand that others may be more used to ``ls`` command.

There is of course no need to completely take way the GNU style flags - we can always keep both. For example, for several DOS shell commands the options /? and -? are treated as identical. It is good to support both so that different users can use the options according to their preferences, and the implementation can be done easily.

Also, note that LS command may not very well with commands like LESS when there are color codes, due to the way that redirection works. This happens regardless of the command shell, such as the internal shell, or the command shells from MS-DOS or FreeDOS. This will become very noticeable when many directory entries are colored, and fixing the LS command itself appears to be the only solution for now.

**Comment by kcgen** [comment]
All sounds good! 
Great point regarding coloured output and the impact on downstream stdin sinks. 

On Linux, most CLI tools use a couple simple  boolean library calls to determine if their output is going to a pipe versus and interactive terminal, and then adjust colorization to suit. (https://unix.stackexchange.com/a/515781)

I suppose on the DOSBox side, when everything is visible under our own shell we can provide such information internally.  But I wonder if the real command.com provided a handler for programs to discover this information?

**Comment by Wengier** [comment]
@kcgen When we use the internal command shell and internal implementation of LS command, it is relatively easy to fix them (I did have sample fixes in the X code). But when the code involves outside command shells or command implementations, it becomes much harder. I have tried the command.com from real DOS (the screenshot below using FreeCOM) and downloaded LESS command, and the output is something like the following:

![image](https://user-images.githubusercontent.com/8216923/148875004-2bba1c2a-bc55-42ba-a29f-c28fca6c2212.png)

**Comment by Grounded0** [comment]
LESS command doesn't belong in a DOS anyway. Anybody that pipes in a DOS probably knows this. Wouldn't spend time on a command that doesn't belong to the official FreeDOS Base Installation. LESS is an external third party tool included in a more comprehensive FreeDOS Full installation. Solution might be to redirect all LESS pipes into MORE.

**Comment by Wengier** [comment]
Strictly speaking neither LS nor LESS is a standard DOS command, but they have been ported to DOS as part of DJGPP for a long time. FreeDOS later bundles them anyway, but they can be used in any DOS, not just FreeDOS. While there may not be necessary to add more non-standard DOS commands, whether a command belongs to the FreeDOS base installation has no direct connections with whether it should be included in our project. The LS command for example is not a FreeDOS base command but has included for some time.

**Comment by Grounded0** [comment]
Well it can be included as LESS redirecting into MORE or writing our internal LESS that works better. Personally i'm not a fan of duplicating Unix commands into DOS because Unix is Unix and DOS is DOS and users have to realise this. Even pipes are not native to DOS but they're being offered.

**Comment by Grounded0** [comment]
Using a LESS command that is open source from some other project could be helpful. If DJGPP LESS works better maybe we should look into that if we cannot just redirect LESS into MORE which works well enough for most. It's a rabbit hole.

**Comment by kcgen** [comment]
Let's stay away from rabbit holes! 🐇

**Comment by MasterO2** [comment]
Now that Staging has the ability to pipe, (https://github.com/dosbox-staging/dosbox-staging/commit/cca0084449f16d297d45ed6638f3c5134f743b9f), what does Staging want to do about improving Staging's implementation of the LS command?

**Comment by Wengier** [comment]
I implemented MORE command and fix for LS command in PR #1556.

**Comment by Burrito78** [comment]
Road to nowhere, here i come. Closed for now during spring cleaning.
